lib/active_merchant/billing/gateways/cyber_source_rest.rb

```diff
@@ -245,68 +320,103 @@ def add_merchant_description(post, options)
         )
       rescue ActiveMerchant::ResponseError => e
         response = e.response.body.present? ? parse(e.response.body) : { 'response' => { 'rmsg' => e.response.msg } }
         Response.new(false, response.dig('response', 'rmsg'), response, test: test?)
       end
 
-      def success_from(response)
-        %w(AUTHORIZED PENDING REVERSED).include?(response['status'])
+      def success_from(action, response, http_method)
+        return response.empty? if http_method == :delete
+
+        case action
+        when /payments/
+          %w(AUTHORIZED PENDING REVERSED).include?(response['status'])
+        else
+          return response['id'].present? && !response['errorInformation'].present? unless http_method == :delete
+        end
       end
 
-      def message_from(response)
-        return response['status'] if success_from(response)
+      def message_from(response, succeeded, http_method)
+        return 'OK' if succeeded && http_method == :delete
+        return response['status'] if succeeded
 
         response['errorInformation']['message']
       end
 
-      def authorization_from(response)
+      def authorization_from(response, action)
         id = response['id']
         has_amount = response['orderInformation'] && response['orderInformation']['amountDetails'] && response['orderInformation']['amountDetails']['authorizedAmount']
         amount = response['orderInformation']['amountDetails']['authorizedAmount'].delete('.') if has_amount
-
+        if /payment-instruments/.match(action) && !response.empty?
+          customer_id = response['_links']['customer']['href'].split('/').last
+          payment_instrument_id = response['id']
+          return [customer_id, payment_instrument_id].join('|')
+        end
         return id if amount.blank?
 
         [id, amount].join('|')
       end
 
-      def error_code_from(response)
-        response['errorInformation']['reason'] unless success_from(response)
+      def error_code_from(response, succeeded)
+        response['errorInformation']['reason'] unless succeeded
       end
 
       # This implementation follows the Cybersource guide on how create the request signature, see:
       # https://developer.cybersource.com/docs/cybs/en-us/payments/developer/all/rest/payments/GenerateHeader/httpSignatureAuthentication.html
       def get_http_signature(resource, digest, http_method = 'post', gmtdatetime = Time.now.httpdate)
-        string_to_sign = {
+        if /customers/.match? resource
+          target = "/tms/v2/#{resource}"
+        elsif %w(instrumentidentifiers paymentinstruments).include? resource
+          target = "/tms/v1/#{resource}"
+        else
+          target = "/pts/v2/#{resource}"
+        end
+
+        hash_to_sign = {
           host: host,
           date: gmtdatetime,
-          "(request-target)": "#{http_method} /pts/v2/#{resource}",
+          "(request-target)": "#{http_method} #{target}",
```

Comment:
💬 Comment:

If we need to build this dynamically is better to rely on the actual URL that will be used and extract the path from there, so we prevent some hard-to-debug issues in the future with the signature.

Reply:
Yes